data/yast2/firstboot/firstboot_hostname.xml

```diff
@@ -0,0 +1,188 @@
+<?xml version="1.0"?>
+<productDefines  xmlns="http://www.suse.com/1.0/yast2ns"
+    xmlns:config="http://www.suse.com/1.0/configns">
+
+    <!--
+    $Id$
+    Work around for the text domain
+    textdomain="firstboot"
+    -->
+
+    <textdomain>firstboot</textdomain>
+
+	<!---
+	See https://github.com/yast/yast-installation/blob/master/doc/control-file.md for more
+	explanation about the control file settings.
+	-->
+
+    <globals>
+
+	<!--
+	If a variable root_password_as_first_user is present in globals section,
+	inst_user step will have the check box
+	    "Use this password for system administrator"
+	so you don't need to include root password step (fate#306297).
+	If the variable is missing (commented), the check box won't appear.
+
+	The value of the variable (true/false) will set the default value for the check box.
+	<root_password_as_first_user config:type="boolean">true</root_password_as_first_user>
+	 -->
+
+	<!-- The default value of "Automatic Login" checkbox -->
+	<enable_autologin config:type="boolean">false</enable_autologin>
+
+	<!-- This option is deprecated in favor of installation-layout -->
+	<!-- <installation_ui>sidebar</installation_ui> -->
+
+	<!-- Configuration of the installation/firstboot layout -->
+	<installation_layout>
+		<mode>steps</mode>
+		<banner config:type="boolean">false</banner>
+	</installation_layout>
+
+	<!--
+	For more variables that can be in this section, look into the control file
+	(/etc/YaST2/control.xml or root directory of installation media).
+	-->
+    </globals>
+    <proposals config:type="list">
+        <proposal>
+            <name>firstboot_hardware</name>
+            <mode>installation</mode>
+            <stage>firstboot</stage>
+            <label>Hardware Configuration</label>
+            <proposal_modules config:type="list">
+                <proposal_module>printer</proposal_module>
+            </proposal_modules>
+        </proposal>
+    </proposals>
+    <workflows  config:type="list">
+        <workflow>
+            <defaults>
+                <enable_back>yes</enable_back>
+                <enable_next>yes</enable_next>
+                <archs>all</archs>
+            </defaults>
+            <stage>firstboot</stage>
+            <label>Configuration</label>
+            <mode>installation</mode>
+            <modules  config:type="list">
+                <module>
+                    <label>Network Autosetup</label>
+                    <name>firstboot_setup_dhcp</name>
+                </module>
+                <module>
+                    <label>Language and Keyboard</label>
+                    <enabled config:type="boolean">true</enabled>
+		    <!-- step for configuration of both language and keyboard layout (fate#306296) -->
+                    <name>firstboot_language_keyboard</name>
+                </module>
+                <module>
+                    <label>Language</label>
+                    <enabled config:type="boolean">false</enabled>
```

Comment:
instead of storing the files, did you consider to edit the xml, with library or some command line. Both files are exactly the same except for those flags.

Reply:
That was my first idea, as indeed the file may get outdated. I wanted to use sed, but for sed at least we can see that the name tag is not always after the "enable" tag. Also some clients do not have the "enable" tag. I will try to think of some approach with XML::Simple.

Reply:
That's tricky. problem with XML::Simple is that it's good for parsing, not really for editing, at best it will produce a different XML file, and that will not be simple anyway. I looked at XML::twig, but as far as I understand the data structure of this xml does not allow easy parsing: "enabled" is not a child node of "Language", but a child of "module"... So I see how to replace enable all client modules ("true" everywhere) but I have no idea how to enable just one specifically. Suggestions welcome...

Reply:
Might be to simple library and I think it doesn't work with namespaces. Another option is to use this library that we already use in the repo and a bit of xpath. Check this:

```perl
use strict;
use warnings;

use XML::LibXML;

my $xml_file = 'firstboot.xml';

my $parser = XML::LibXML->new();
my $doc = $parser->parse_file($xml_file);

my $xpc = XML::LibXML::XPathContext->new($doc);
$xpc->registerNs('ns', 'http://www.suse.com/1.0/yast2ns');

my ($node_to_edit) = $xpc->findnodes("/ns:productDefines/ns:workflows/ns:workflow/ns:modules/ns:module[./ns:label[text()='Host Name']]/ns:enabled");
$node_to_edit->removeChildNodes();
$node_to_edit->appendText('true');
$node_to_edit->to_literal;
#print $node_to_edit;
print $doc->toString;
```

Reply:
That's great, thanks. I tried to apply it, however, it's still tricky as we need to edit a file that's on the SUT... I tried to parse the data with parse_chunk, parse_string, parse_balanced_chunk, but all gave me errors, so I end up creating a temporary file on the worker, which works, but, well...

Reply:
Also, because we have to do some echo of an entire xml file, the module now lasts over 4 minutes, when it was just 26s before :(

Reply:
you can use the library just passing the content with `cat` using openqa function to interact with the SUT.

Reply:
According with disscussion offline in RC, it is better the way you are currently trying with static files, because we would introduce too much complexity moving the file from one place to another in a proper way and it is not even as customer would do. So the only good options would be (1) to have it in static way as you already trying and (2) find some fancy text replacement. And (2) is not really good idea as you would have to hardcode some tabs, etc. Because the files is not likely to change and we would have there to compare if it breaks in customers, then I would suggest to continue with having the file static as you initially did. Hope you have some fun with the library trying out (it could be useful in other task for sure) despite it wasn't a good suggestion.

Reply:
Another option to avoid those hardcode tab is install some dep, I tried this tool and it is pretty cool and available in sle server repo:
`xmlstarlet ed -L -N ns=http://www.suse.com/1.0/yast2ns -u "/ns:productDefines/ns:workflows/ns:workflow/ns:modules/ns:module[./ns:label[text()='Host Name']]/ns:enabled" -v "true" ./firstboot.xml`

Reply:
That looks like the best option to me. The control file being part of the firstboot package, it is subject to change and is actually different currently in TW vs SLE. So, awesome finding ! Never heard of that tool before.

Reply:
In the end, another option was chosen, but thanks anyway for the great support :)